core/src/main/java/org/apache/iceberg/OverwriteData.java

```diff
@@ -88,6 +119,42 @@ public OverwriteFiles validateAddedFiles() {
       }
     }
 
+    if (conflictDetectionFilter != null) {
+      PartitionSpec spec = writeSpec();
+      Expression inclusiveExpr = Projections.inclusive(spec).project(conflictDetectionFilter);
+      Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+
+      InclusiveMetricsEvaluator metrics = new InclusiveMetricsEvaluator(base.schema(), conflictDetectionFilter);
+
+      List<DataFile> newFiles = collectNewFiles(base);
+      for (DataFile newFile : newFiles) {
+        ValidationException.check(
+            !inclusive.eval(newFile.partition()) || !metrics.eval(newFile),
+            "A conflicting file was appended that matches filter '%s': %s",
+            conflictDetectionFilter, newFile.path());
+      }
+    }
+
     return super.apply(base);
   }
+
+  private List<DataFile> collectNewFiles(TableMetadata meta) {
+    List<DataFile> newFiles = new ArrayList<>();
+
+    Long currentSnapshotId = meta.currentSnapshot() == null ? null : meta.currentSnapshot().snapshotId();
```

Comment:
I think that if the current snapshot is null, the loop is skipped and this returns an empty list. That's not correct, right?

Reply:
Not sure I got. If the current snapshot is null, it means the table is still empty and shouldn't have any new files, right?